client/src/entrypoints/admin/userbar.js

```diff
@@ -49,10 +51,120 @@ document.addEventListener('DOMContentLoaded', (e) => {
+    if (userbar.classList.contains(isActiveClass)) {
       hideUserbar();
     } else {
-      showUserbar();
+      showUserbar(true);
+    }
+  }
+
+  function setFocusToTrigger() {
+    setTimeout(() => trigger.focus(), 300);
+  }
+
+  function isFocusOnItems() {
+    let isFocused = false;
+    list.querySelectorAll('a').forEach((element) => {
+      if (element === document.activeElement) {
+        isFocused = true;
+      }
+    });
+    return isFocused;
+  }
+
+  function setFocusToFirstItem() {
+    if (listItems.length > 0) {
+      setTimeout(() => {
+        listItems[0].firstElementChild.focus();
+      }, 100); // Workaround for focus bug
+    }
+  }
+
+  function setFocusToLastItem() {
+    if (listItems.length > 0) {
+      setTimeout(() => {
+        listItems[listItems.length - 1].firstElementChild.focus();
+      }, 100); // Workaround for focus bug
+    }
+  }
+
+  function setFocusToNextItem() {
+    listItems.forEach((element, idx) => {
+      // Check which item is currently focused
+      if (element.firstElementChild === document.activeElement) {
+        setTimeout(() => {
+          if (idx + 1 < listItems.length) {
+            // Focus the next item
+            listItems[idx + 1].firstElementChild.focus();
+          }
+        }, 100); // Workaround for focus bug
+      }
+    });
+  }
+
+  function setFocusToPreviousItem() {
+    // Check which item is currently focused
+    listItems.forEach((element, idx) => {
+      if (element.firstElementChild === document.activeElement) {
+        setTimeout(() => {
+          if (idx > 0) {
+            // Focus the previous item
+            listItems[idx - 1].firstElementChild.focus();
+          }
+        }, 100); // Workaround for focus bug
+      }
+    });
+  }
+
+  function handleKeyDown(event) {
+    // Only handle keyboard input if the userbar is open
+    if (userbar.classList.contains(isActiveClass)) {
+      if (event.key === 'Escape') {
+        hideUserbar();
+        setFocusToTrigger();
+        return;
+      }
+
+      if (isFocusOnItems()) {
+        switch (event.key) {
+        case 'ArrowDown':
+          event.preventDefault();
+          setFocusToNextItem();
+          break;
+        case 'ArrowUp':
+          event.preventDefault();
+          setFocusToPreviousItem();
+          break;
+        case 'Home':
+          event.preventDefault();
+          setFocusToFirstItem();
+          break;
+        case 'End':
+          event.preventDefault();
+          setFocusToLastItem();
+          break;
+        default:
+          break;
+        }
+      }
+      return;
+    }
+    // Check if the userbar is focused (but not open yet) and should be opened by keyboard input
+    if (trigger === document.activeElement) {
+      switch (event.key) {
+      case 'ArrowUp':
+        event.preventDefault();
+        showUserbar(false);
+        setTimeout(() => setFocusToFirstItem(), 300); // Workaround for focus bug
```

Comment:
I wasn't sure why I put it there in the first place 😅

Turns out that I put this here for additional delay in shifting focus. The userbar opening animation interferes and without the additional delay it will not properly focus.

The duration of the animation depends on the amount of items in the userbar, meaning that if there are enough items 300ms might not be enough. But this is an unlikely scenario imo as each item only adds .05s to the animation duration.

I've added additional comments in f113ff883613dd7d992923d6cb93367bb9390e2b to explain why this code is here. Marking as resolved.